File: ui/visual-text-editor/src/main/java/org/visual/text/editor/GenericTextArea.java
package org.visual.text.editor;

import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.css.PseudoClass;
import javafx.scene.layout.Region;

public class GenericTextArea extends Region {
  private final PseudoClass READ_ONLY = PseudoClass.getPseudoClass("readonly");
  private final PseudoClass HAS_CARET = PseudoClass.getPseudoClass("has-caret");
  private final PseudoClass FIRST_PAR = PseudoClass.getPseudoClass("first-paragraph");
  private final PseudoClass LAST_PAR = PseudoClass.getPseudoClass("last-paragraph");

  private final BooleanProperty editable =
      new SimpleBooleanProperty(this, "editable", true) {
        @Override
        protected void invalidated() {
          ((Region) getBean()).pseudoClassStateChanged(READ_ONLY, !get());
        }
      };
}

File: ui/visual-model-text-editor/src/main/kotlin/org/visual/model/text/editor/GenericTextArea.kt
package org.visual.model.text.editor

import java.util.*
import javafx.beans.property.BooleanProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.css.PseudoClass
import javafx.scene.layout.Region

class GenericTextArea : Region() {
  private val READ_ONLY: PseudoClass = PseudoClass.getPseudoClass("readonly")
  private val HAS_CARET: PseudoClass = PseudoClass.getPseudoClass("has-caret")
  private val FIRST_PAR: PseudoClass = PseudoClass.getPseudoClass("first-paragraph")
  private val LAST_PAR: PseudoClass = PseudoClass.getPseudoClass("last-paragraph")
  private val editable: BooleanProperty =
      object : SimpleBooleanProperty(this, "editable", true) {
        override fun invalidated() {
          (bean as Region).pseudoClassStateChanged(READ_ONLY, !get())
        }
      }
}
